### PR TITLE
HDDS-7812. Recon UI: Add visualization for container size distribution.

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?
Recon has an API endpoint to get container size distribution in Ozone. Add visualisation in Recon UI for this using histograms.
We will have to add another box that will display the percentages
Example
```
    {      
     "containerSize": 1024,       
      "count": 2     
    },
    {
      "containerSize": 16384,
      "count": 3
    },
    {
      "containerSize": 2097152,
      "count": 4
    },
    {
      "containerSize": 11270,
      "count": 5
    }
```
Total Count ➖ 2+3+4+5 = *14 Total Containers*
Percentage of Containers having upperbound 1024 = 2/14 *100 = 14.28%
Percentage of Containers having upperbound 16384 = 3/14 *100 = 21.42%
Percentage of Containers having upperbound 2097152 = 4/14 *100 = 28.5%
Percentage of Containers having upperbound 11270 = 5/14 *100 = 35.7%

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7812

## How was this patch tested?
Manually
image

image

For Empty Response 

image
------------------------------------
For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  }
]
```

image
---------------------------------------------------------------------------------------------------------------------------------------------

For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 1
  }
]
```

image
----------------------------------------------------------------------------------------------------------------------------------------------
For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 5
  }
]
```
image
------------------------------------------------------------------------------------------------------------------------------------------------------
For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 10
  }
]
```
image
---------------------------------------------------------------------------------------------------------------------------------------------------
Sample Data for Testing
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 19
  },
  {
    "containerSize": 5306870912,
    "count": 5
  },
  {
    "containerSize": 10073741824,
    "count": 20 
  },
  {
    "containerSize": 100073741824,
    "count": 14
  }
]
```
Color 
Blue :- Count 20
Orange  :- Count 19
Green :- Count 14
Red :- Count 5
Violet :- Count 1

image
---------------------------------------------------------------------------------
Sample Data for Testing
```
[
  {
    "containerSize": 5306870912,
    "count": 5
  },
  {
    "containerSize": 10073741824,
    "count": 20
  },
  {
    "containerSize": 1000073741824,
    "count": 14
  }
]
```
Color 
Blue :- Count 20
Green  :- Count  5
Orange :- Count 14

image
---------------------------------------------------------

### CODE REVIEW

**Comment by ArafatKhan2198** [comment]
@smitajoshi12 Can you please add the screenshot of the latest Pie Chart that we had discussed !!

**Comment by smitajoshi12** [comment]
@ArafatKhan2198  These are latest changes of patch.

**Comment by ArafatKhan2198** [comment]
@smitajoshi12 These screenshots indicate that the size of some containers is less than 512 MB, which is not possible according to the backend. To address this, I implemented changes to the backend that group all containers with sizes smaller than 512 MB into the range of 0-512 MB. Please pull in the latest changes and populate a local ozone cluster with data and take scereenshots of those.

**Comment by smitajoshi12** [comment]
> @smitajoshi12 These screenshots indicate that the size of some containers is less than 512 MB, which is not possible according to the backend. To address this, I implemented changes to the backend that group all containers with sizes smaller than 512 MB into the range of 0-512 MB. Please pull in the latest changes and populate a local ozone cluster with data and take scereenshots of those.

@ArafatKhan2198 
Added new screenshots.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 1]
```diff
@@ -36,10 +37,17 @@ interface IFileCountResponse {
...
+  containerSize: number;
+  count: number;
+}
+
 interface IInsightsState {
   isLoading: boolean;
   fileCountsResponse: IFileCountResponse[];
+  containerCountResponse: IContainerCountResponse[];
   plotData: Plotly.Data[];
+  containerData: Plotly.Data[];
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:50]: as we are having both container and file count size task now, we should rename the previous variable, so it is easier to understand it. 
```suggestion
  fileCountData: Plotly.Data[];
  containerCountData: Plotly.Data[];
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 2]
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
         const lowerbound = upperboundPower > 10 ? size(2 ** (upperboundPower - 1)) : size(0);
         return `${lowerbound} - ${upperbound}`;
       });
+      
+      const xyConatainerCountMap: Map<number, number> = containerCountResponse.reduce(
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:169]:
```suggestion
      const xyContainerCountMap: Map<number, number> = containerCountResponse.reduce(
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 3]
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
...
+        (map: Map<number, number>, current) => {
+          const fileSize = current.containerSize;
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:171]:
```suggestion
          const containerSize = current.containerSize;
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 4]
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
...
+      let keysize = [];
+      keysize = Array.from(xyConatainerCountMap.keys()).map(value => {
+        return (size(value) );
+      });
+
       this.setState({
         plotData: [{
           type: 'bar',
           x: xValues,
           y: Array.from(xyMap.values()),
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:196]:
```suggestion
          y: Array.from(xyFileCountMap.values()),
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 5]
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
...
+
+      let keysize = [];
+      keysize = Array.from(xyConatainerCountMap.keys()).map(value => {
+        return (size(value) );
+      });
+
       this.setState({
         plotData: [{
           type: 'bar',
           x: xValues,
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:195]:
```suggestion
          x: xFileCountValues,
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 6]
```diff
@@ -215,69 +264,107 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
...
+                                    size: 15
+                                  },
+                                  title: 'Container Size Distribution',
+                                  showlegend: true
+                                }
+                              }/>
+                          </Col>
+                        </Row>
+                      </div> :
+                      <div>No data to visualize Container size distribution. Add files to Ozone to see a visualization on Container size distribution.</div>)}
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:363]:
```suggestion
                      <div>No data to visualize container size distribution. Add files to Ozone to see a visualization on container size distribution.</div>)}
```

**Review by dombizita — COMMENTED** [review]
thanks for working on this @smitajoshi12, I added a few comments inline regarding naming, most of them are because now we have both file and container size distribution tasks, so the variables should be named inline with what are they used for, please take a look at my suggestions

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx (hunk 1 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx]: @dombizita 
Completed Change Suggested by you.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx (hunk 3 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx]: @dombizita 
Completed Change Suggested by you.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx (hunk 4 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx]: @dombizita 
Completed Change Suggested by you.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx (hunk 6 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx]: @dombizita 
Completed Change Suggested by you.

**Comment by dombizita** [comment]
can you also review this patch @ArafatKhan2198?

**Comment by ArafatKhan2198** [comment]
Can you please make the allingment proper, the some of the letters get hidden :-
image

**Comment by ArafatKhan2198** [comment]
Please ensure that the text color is white for those specific colors where it currently appears as black.

image

**Comment by ArafatKhan2198** [comment]
Change the message to ➖ ”No data available for container size distribution visualization. Add files to Ozone.”
which sounds more simple and short

image

**Comment by smitajoshi12** [comment]
> Change the message to heavy_minus_sign ”No data available for container size distribution visualization. Add files to Ozone.” which sounds more simple and short
> 
> image



> Can you please make the alinement proper, some of the letters get hidden :- image

Completed Changes Suggested by Arafat.

**Comment by smitajoshi12** [comment]
> Change the message to heavy_minus_sign ”No data available for container size distribution visualization. Add files to Ozone.” which sounds more simple and short
> 
> image

Completed Changes Suggested by Arafat.

image

**Comment by smitajoshi12** [comment]
> Please ensure that the text color is white for those specific colors where it currently appears as black.
> 
> image

@ArafatKhan2198 
In Disk Usage it is internally manged by library we could not modify color of tool-tip as per background color it will vary.

**Comment by smengcl** [comment]
Thanks @smitajoshi12 for the patch. Thanks @dombizita for reviewing this.

**Pull request merged**